normandy/selfrepair/static/js/self_repair_driver.js

```diff
@@ -0,0 +1,132 @@
+                reject(req);
+            } else {
+                resolve(req);
+            }
+        });
+        req.open(method, url);
+
+        let data = undefined;
+        if (options.data) {
+            req.setRequestHeader('Content-Type', 'application/json');
+            data = JSON.stringify(data);
+        }
+
+        if (options.headers) {
+            for (let key in options.headers) {
+                req.setRequestHeader(key, options.headers[key]);
+            }
+        }
+
+        req.send(data);
+    });
+}
+xhr.get = xhr.bind(null, 'GET');
+xhr.post = xhr.bind(null, 'POST');
+
+
+let registeredActions = {};
+window.Normandy = {
+    registerAction(name, func) {
+        registeredActions[name] = func;
+    }
+};
+
+
+/**
+ * Download the implementation of the given action from the server.
+ *
+ * @param {Action} action  - Action taken from a Recipe.
+ * @promise Resolves once the action has been loaded and registered.
+ * @rejects {Error} Rejects if the action could not be loaded or did not
+ *     register itself.
+ */
+function loadAction(action) {
+    return new Promise((resolve, reject) => {
+        if (!registeredActions[action.name]) {
+            let script = document.createElement('script');
+            script.src = action.implementation.url;
+            script.onload = () => {
+                if (!registeredActions[action.name]) {
+                    reject(new Error(`Could not find action with name ${action.name}.`));
+                } else {
+                    resolve();
+                }
+            };
+            document.head.appendChild(script);
+        } else {
+            resolve();
+        }
+    });
+}
+
+/**
+ * Fetch recipes from the Recipe server.
+ *
+ * @promise {Array<Recipe>} List of recipes.
+ */
+function fetchRecipes() {
+    let {recipe_url, locale} = document.documentElement.dataset;
+
+    return xhr.post(recipe_url, {
+        data: {locale: locale},
+        headers: {Accept: 'application/json'}
+    }).then(request => {
+        return JSON.parse(request.responseText).recipes;
+    });
+}
+
+
+/**
+ * Fetch and execute the actions for the given recipe.
+ *
+ * @param {Recipe} recipe - Recipe retrieved from the server.
+ * @promise Resolves once all actions are complete.
+ */
+function runRecipe(recipe) {
+    return Promise.all(recipe.actions.map(loadAction)).then(() => {
+        let recipePromise = Promise.resolve();
+
+        for (let action of recipe.actions) {
+            let func = registeredActions[action.name];
+            recipePromise = recipePromise.then(() => {
+                return func(null, action.arguments);
+            });
+        }
+
+        return recipePromise;
+    });
+}
+
+
+// Actually fetch and run the recipes.
+fetchRecipes().then((recipes) => {
+    let chain = Promise.resolve();
+
+    for (let recipe of recipes) {
+        chain.then(() => {
+            return runRecipe(recipe);
+        });
+    }
+
+    return chain;
```

Comment:
I meant like maybe we make a helper function that removes it:

```js
promiseChain(list_of_promises);
```

Initially I don't think that'd work because by the time you've generated the promises, they're already executing, but maybe we can figure something out. Maybe it's not worth the time. I'll think about it.